efi/fw_load_handler.go

```diff
@@ -236,22 +237,90 @@ func (h *fwLoadHandler) measureBootManagerCodePreOS(ctx pcrBranchContext) error
+	events := h.log.Events
+	measuredSeparator := false
+	for len(events) > 0 {
+		event := events[0]
+		events = events[1:]
+
 		if event.PCRIndex != tcglog.PCRIndex(bootManagerCodePCR) {
 			continue
 		}
 
 		if event.EventType == tcglog.EventTypeSeparator {
-			return h.measureSeparator(ctx, bootManagerCodePCR, event)
+			if err := h.measureSeparator(ctx, bootManagerCodePCR, event); err != nil {
+				return err
+			}
+			measuredSeparator = true
+			break
 		}
 		ctx.ExtendPCR(bootManagerCodePCR, tpm2.Digest(event.Digests[ctx.PCRAlg()]))
 	}
 
-	return errors.New("missing separator")
+	if !measuredSeparator {
+		return errors.New("missing separator")
+	}
+
+	// Some newer laptops including those from Dell and Lenovo execute code from a firmware volume as part
+	// of the OS-present environment, before shim runs, and using the LoadImage API which results in an
+	// additional measurement to PCR4. Copy this into the profile if it's part of a well-known endpoint
+	// management application, else return an error. Anything else here will be picked up by the pre-install
+	// checks.
+	for len(events) > 0 {
+		event := events[0]
+		events = events[1:]
+
+		if event.PCRIndex != tcglog.PCRIndex(bootManagerCodePCR) {
+			continue
+		}
+		if event.EventType != tcglog.EventTypeEFIBootServicesApplication {
+			return fmt.Errorf("unexpected OS-present event type: %v", event.EventType)
+		}
+
+		data, ok := event.Data.(*tcglog.EFIImageLoadEvent)
+		if !ok {
+			return fmt.Errorf("invalid event data for OS-present event: %w", event.Data.(error))
+		}
+
+		if len(data.DevicePath) == 0 {
+			return errors.New("invalid device path for first OS-present image load event: device path is empty")
+		}
+		if _, isFv := data.DevicePath[0].(efi.MediaFvDevicePathNode); !isFv {
+			// Not loaded from flash, so we'll break here, assuming that this is the first OS component
+			break
+		}
+
+		// The image is loaded from flash - we should have a path of the form "Fv()\FvFile()".
+		if len(data.DevicePath) != 2 {
+			return fmt.Errorf("invalid firmware volume device path (%v) for OS-present image load: invalid length", data.DevicePath)
+		}
+		fvf, isFvf := data.DevicePath[1].(efi.MediaFvFileDevicePathNode)
+		if !isFvf {
+			// The second component should be the firmware volume file name.
+			return fmt.Errorf("invalid firmware volume device path (%v) for OS-present image load: doesn't terminate with FvFile", data.DevicePath)
+		}
+
+		name, known := guids.FileNameString(efi.GUID(fvf))
+		if !known {
+			return fmt.Errorf("unknown image loaded from firmware volume during OS-present: %v", data.DevicePath)
+		}
+		switch name {
+		case "AbsoluteAbtInstaller", "AbsoluteComputraceInstaller":
+			ctx.ExtendPCR(bootManagerCodePCR, tpm2.Digest(event.Digests[ctx.PCRAlg()]))
+		default:
+			return fmt.Errorf("unexpected image loaded from firmware volume during OS-present: %v", data.DevicePath)
```

Comment:
we would hit the errors here already when installing, right?

Reply:
Yes, in fact the check in my current branch uses a shared function `isAbsoluteAgentLaunch(*tcglog.Event)` - I was hoping to push an update to this before you got to reviewing it that would introduce the shared function. I'm currently in the process of splitting the new checks into a `efi/preinstall` sub-package that will have a main `RunChecks` API, but untangling things is a bit of a challenge (not impossible) and there'll need to be another internal package with a few helpers that's shared by both packages.